cpp/src/rolling/jit/code/kernel.cpp

```diff
@@ -32,27 +32,134 @@ const char* kernel =
 R"***(
 #include "operation.h"
 
-template <typename WindowType>
-cudf::size_type get_window(WindowType window, cudf::size_type index) { return 0; }
+template<class T>
+T minimum(T a, T b){
+  return a < b ? a : b;
+}
 
-template <>
-cudf::size_type get_window(cudf::size_type window, cudf::size_type index) { return window; }
+struct preceding_window_wrapper {
 
-template <>
-cudf::size_type get_window(cudf::size_type* window, cudf::size_type index) {
-  return window[index];
+  const cudf::size_type *d_group_offsets;
+  const cudf::size_type *d_group_labels;
+  cudf::size_type window;
+  
+  cudf::size_type operator[](cudf::size_type idx){
+    auto group_label = d_group_labels[idx];
+    auto group_start = d_group_offsets[group_label];
+    return minimum(window, idx - group_start + 1); // Preceding includes current row.
+  }
+
+};
+
+struct following_window_wrapper {
+
+  const cudf::size_type *d_group_offsets;
+  const cudf::size_type *d_group_labels;
+  cudf::size_type window;
+  
+  cudf::size_type operator[](cudf::size_type idx){
+    auto group_label = d_group_labels[idx];
+    auto group_end = d_group_offsets[group_label+1]; // Cannot fall off the end, since offsets is capped with `input.size()`.
+    return minimum(window, (group_end - 1) - idx);
+  }
+
+};
+
+template<class Iter, class T>
+cudf::size_type __device__ upper_bound(Iter iter, cudf::size_type first, cudf::size_type last, const T& value)
+{
+    cudf::size_type current;
+    cudf::size_type count, step;
+    count = last - first;
+
+    while (count > 0) {
```

Comment:
I have concerns about this being used as an inner loop inside the kernel. What's the theoretical performance impact of that, and can we avoid it, perhaps by pre-computing the lower/upper bounds?